### ISSUE TITLE
Rails: After migration from ruby to jruby, app is not booting

### ISSUE THREAD

**Issue by kushthedude** [jruby/jruby#6310]
**Environment Information**
  
Provide at least:
* JRuby version : 9.2.12.0
* Operating system and platform: MacOS Cataline, rails 6.0.3

Other relevant info you may wish to add:
* Installed or activated gems
* Application/framework version (e.g. Rails, Sinatra)
* Environment variables

**Expected Behavior**
I have migrated my app from ruby to JRuby, I replaced my pg adapter with activerecord-jdbc-adapter & when I start my application using `rails server` & visit the app, the application crashes with error :- 
```sh 
=> Booting Puma
=> Rails 6.0.3.2 application starting in development
=> Run `rails server --help` for more startup options
The signal INFO is in use by the JVM and will not work correctly on this platform
Puma starting in single mode...
* Version 4.3.5 (jruby 9.2.12.0 - ruby 2.5.7), codename: Mysterious Traveller
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://[::1]:8080
* Listening on tcp://127.0.0.1:8080
Use Ctrl-C to stop
   (2.4ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
  User Load (30.5ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? ORDER BY "users"."id" ASC LIMIT ?  [["id", 2], ["LIMIT", 1]]
  ↳ app/controllers/welcome_controller.rb:5:in `home'
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jruby.ext.openssl.SecurityHelper (file:/Users/kush/.rvm/rubies/jruby-9.2.12.0/lib/ruby/stdlib/jopenssl.jar) to field java.security.MessageDigest.provider
WARNING: Please consider reporting this to the maintainers of org.jruby.ext.openssl.SecurityHelper
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
java(19877,0x700008f3b000) malloc: *** error for object 0x7fb3213cf000: pointer being freed was not allocated
java(19877,0x700008f3b000) malloc: *** set a breakpoint in malloc_error_break to debug
[1]    19877 abort      rails server
```

**Actual Behavior**

App should work fine

**Comment by headius** [comment]
Someone else reported something similar on Twitter. There should be a crash report somewhere in the same directory, can you post it as a [gist](https://gist.github.com) and then post the URL here?

**Comment by headius** [comment]
It would also be helpful if you could provide your version of Java, from `java -version`.

**Comment by rebelwarrior** [comment]
I'm the guy from Twitter. Experiencing intermittent fails on my rails app. Coded in JRuby from the get go then after any of a number of updates having this problem. Usually boots but once I load a page I get a malloc error. 

My system
java: 
```
openjdk 13.0.2 2020-01-14
OpenJDK Runtime Environment (build 13.0.2+8)
OpenJDK 64-Bit Server VM (build 13.0.2+8, mixed mode, sharing)
```
OS: MacOS 10.15.5 Catalina, Macbook Air 4 GB of memory
Database: PostgresQL fired w/ Postgress App
Rails: 5.2.4.3
JRuby: 9.2.12.0 & 9.2.11.1 

At first I thought it was the kramdown gem but then I realized it was intermittent and not related to that gem at all. I would get a fail, then restart without changing anything and it would work. However once I had a crash, I was almost guaranteed to have another until a restart. 

Attached more info. 
[bugreport.java.txt](https://github.com/jruby/jruby/files/4893019/bugreport.java.txt)
[updatedGems_changed.txt](https://github.com/jruby/jruby/files/4893024/updatedGems_changed.txt)
![Screen Shot 2020-07-07 at 4 21 23 PM](https://user-images.githubusercontent.com/839013/86965957-549a1400-c136-11ea-89be-80623c1c1184.png)

**Comment by rebelwarrior** [comment]
Gist w/ Crash report 

[](https://gist.github.com/rebelwarrior/748fefff4d8682fffa6ab12550ae3b92)

**Comment by headius** [comment]
@rebelwarrior Thank you for that information!

After looking at the crash report, this does *not* look like it's a JRuby issue... or at least not directly.

The thread that crashes is part of the G1 garbage collector. This may indicate a bug in G1. It's also possible we're overwriting some memory that G1 manages, but I'd expect to see more unpredictable crashes (i.e. it should crash rather differently every time).

You could try to confirm if this is a G1 problem by switching to a different garbage collector. Java 13 ships a couple other collectors:

* The "parallel" collector, which you can enable using `-XX:+UseParallelGC` JVM flag (with `-J` prefix if passed to JRuby).
* Shenandoah, which I believe is in most (all?) OpenJDK builds. Article linked below has more information.
* ZGC, which might be in most recent OpenJDK builds, but it still considered experimental. Article linked below has more information.

Article in question: https://blogs.oracle.com/javamagazine/understanding-the-jdks-new-superfast-garbage-collectors

You might also try using Java 14, or making sure you're on the latest build of Java 13.

**Comment by rebelwarrior** [comment]
I will upgrade to java 14 first. Thank you!!

**Comment by kushthedude** [comment]
Hey @headius ,
I am on the latest build of java 14
```sh
java 14.0.1 2020-04-14
Java(TM) SE Runtime Environment (build 14.0.1+7)
Java HotSpot(TM) 64-Bit Server VM (build 14.0.1+7, mixed mode, sharing)
```
But same error 
```sh
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jruby.ext.openssl.SecurityHelper (file:/Users/kush/.rvm/rubies/jruby-9.2.12.0/lib/ruby/stdlib/jopenssl.jar) to field java.security.MessageDigest.provider
WARNING: Please consider reporting this to the maintainers of org.jruby.ext.openssl.SecurityHelper
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
java(12812,0x7000047a8000) malloc: *** error for object 0x7fb6cfbb2460: pointer being freed was not allocated
java(12812,0x7000047a8000) malloc: *** set a breakpoint in malloc_error_break to debug
[1]    12812 abort      rails s
```

**Comment by rebelwarrior** [comment]
More errors w/ openjdk 14
![Screen Shot 2020-07-10 at 2 04 28 PM](https://user-images.githubusercontent.com/839013/87184634-5be82b80-c2b6-11ea-8575-11b37cc30595.png)

[https://gist.github.com/rebelwarrior/d7231ef2f5c9d1dd86be37f9205e4630](https://gist.github.com/rebelwarrior/d7231ef2f5c9d1dd86be37f9205e4630)

**Comment by rebelwarrior** [comment]
FYI I'm using `bin/rails s` to launch it. Don't know if that has anything to do with it. What's the command to rebuild that.

**Comment by kares** [comment]
as noted above, could you guys confirm whether it keeps happening with `export JRUBY_OPTS="-J-XX:+UseParallelGC"`

**Comment by rebelwarrior** [comment]
Will try that next. Thanks for posting the whole command.

**Comment by rebelwarrior** [comment]
@kares Tried with the export on my zshrc before running the app. Had a crash anyway. 
![Screen Shot 2020-07-13 at 9 30 36 PM](https://user-images.githubusercontent.com/839013/87370125-22325180-c550-11ea-81bb-f190ef6b09ca.png)

This one did not produce a debug log like the previous ones.

**Comment by rebelwarrior** [comment]
Don't know if related but I had a Kernel Panic today as well, unrelated to JRuby (I was running pages and pdf expert). Probably no connection but thought I get a gist of it as well. I removed Virtualbox after this. 
[https://gist.github.com/rebelwarrior/04986ba7eea6b17337cb7850d38fb321](https://gist.github.com/rebelwarrior/04986ba7eea6b17337cb7850d38fb321)

**Comment by rebelwarrior** [comment]
Could be my perception but it seems to be crashing more often with the parallel GC
![Screen Shot 2020-07-14 at 11 51 47 AM](https://user-images.githubusercontent.com/839013/87447418-90b2f600-c5c8-11ea-976d-9de43757d7c0.png)

**Comment by rebelwarrior** [comment]
Got a weird warning when running rake routes. In case it provides some info. [https://gist.github.com/rebelwarrior/a2bfcec303a6c470a448d117bd0e94eb](https://gist.github.com/rebelwarrior/a2bfcec303a6c470a448d117bd0e94eb)

**Comment by headius** [comment]
@rebelwarrior Can you provide a crash report with +ParallelGC enabled?

**Comment by rebelwarrior** [comment]
None was produced last time.

**Comment by headius** [comment]
@rebelwarrior Are you able to identify the last version of JRuby which did not seem to have this problem?

**Comment by rebelwarrior** [comment]
I used the previous version 9.2.1.11 I believe for months without issue. After I upgraded to Catalina and openjdk 13 point something that’s when they started to show up.  

They’re not predictable. I worked most of the day with it running and no issues today. (Back using the regular GC). 

I’ll post any new crash logs I get. 

I’ll switch to the parallel GC tonight and see what happens. 

- David.

**Comment by headius** [comment]
I just found this: https://github.com/sass/sassc-ruby/issues/208

It's possible this crash you have been seeing is from sassc. Downgrading the sass-rails dependency to an earlier version that uses sass-ruby may be a workaround.

I am not sure whether the sassc crash is sassc's fault or JRuby's fault.

**Comment by headius** [comment]
None of you upgrade Rails did you? They changed from sass-ruby to sassc-ruby in a minor 6.0 update: https://github.com/rails/rails/issues/38492

You can imagine my frustration.

**Comment by headius** [comment]
> Got a weird warning when running rake routes. In case it provides some info.

This is not likely related but could indicate some rare bug in the debugging/tracing logic in JRuby. Probably should be filed as a separate issue if you are able to reproduce it.

**Comment by headius** [comment]
I have pushed https://github.com/jruby/jruby/pull/6331 in an attempt to address this, but I have no reliable way to reproduce the issues you are seeing. Help me verify!

**Comment by headius** [comment]
I have confirmation that #6331 appears to fix #6284 so we'll go with it. Should be in a jruby-9.2 snapshot soon and hopefully one of you can try to confirm it in your apps.

**Comment by rebelwarrior** [comment]
May not be 100% related but I think it might. My Heroku deploy of the app above kept crashing on the asset:precompile. I down graded sassc from 2.4.0 to hard set at 2.3.0. Then it went well. Heroku uses openjdk-1.8
